### PR TITLE
Implement translation support

### LINKED ISSUE #66: Translation support
Hi,

Have see patch in FR for diablo but on linux with devilutionx how to make that ?

Best Regards

### PR DESCRIPTION
**Pull request by AJenbo**
This brings translation support to the game, it will probably need a bit of adjustment on different platforms and for the handling of static texts. Currently, also only the strings from DiabloUI and appfat.cpp are marked as translatable.

Fixes #66

Skærmbillede fra 2020-01-07 23-02-49

### CODE REVIEW

File: SourceX/DiabloUI/fonts.cpp [hunk 1]
```diff
@@ -52,15 +52,15 @@ void UnloadArtFonts()
 void LoadTtfFont() {
 	if (!TTF_WasInit()) {
 		if (TTF_Init() == -1) {
-			printf("TTF_Init: %s\n", TTF_GetError());
+			SDL_Log(_("TTF_Init: %s\n"), TTF_GetError());
```
Comment (glebm) [inline SourceX/DiabloUI/fonts.cpp]: There's no need to translate this and most other SDL_Log calls.

**Comment by AJenbo** [comment]
@glebm how do you feel about the way I'm handling the translation of static texts here? I saw it used in other projects, all strings are wrapped in a fake translation function (`N_()`) and the actual translation then happens when the texts are scheduled for rendering: https://github.com/diasurgical/devilutionX/pull/533/commits/1acd1794463cd5f8691e9b30891296a3ee289a45

**Comment by glebm** [comment]
@AJenbo There is a danger with that approach when mixing `N_` and `_` calls, described here: https://www.gnu.org/software/gettext/manual/html_node/Special-cases.html

For UI components, perhaps we could mark it explicitly with a wrapper type:

```c++
struct UiStringView {
  constexpr UiStringView(const char *text, bool localize = false) 
    : localize_(localize), text_(text) {
  }

  const char *localized_text() {
    if (localize_) return gettext(text_);
    return text_;
  }
  // (or make the fields public and move the logic to render/display code)

 private:
  bool localize_;
  const char *text_;
};
```

The macro would return this type and the static arrays would change from:

```diff
-const char *const CREDITS_LINES[] = {
+const UiStringView CREDITS_LINES[] = {
```

The type above is implicitly constructible from a C-string literal.

**Comment by glebm** [comment]
E.g. like this: https://godbolt.org/z/DzaUAy

**Comment by AJenbo** [comment]
billede

**Comment by AJenbo** [comment]
If any one can getting this building on Windows, Mac and Swtich I would appreciate the help.

**Comment by AJenbo** [comment]
Looks like https://github.com/SuperTux/supertux is a good example of how to get this implemented for Windows and Mac

File: translations/da.po [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+msgid "Serial Communications Programming"
+msgstr "Seriel kommunikationsprogrammering"
+
+#: SourceX/DiabloUI/credits_lines.cpp:36
+msgid "Installer Programming"
+msgstr "Installationsprogrammering"
+
+#: SourceX/DiabloUI/credits_lines.cpp:39
+msgid "Art Directors"
+msgstr "Kunst Instruktør"
```
Comment (AJenbo) [inline translations/da.po:204]:
```suggestion
msgstr "Kunst instruktører"
```

File: translations/da.po [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+msgid "Voice Production, Direction & Casting"
+msgstr "Stemmeproduktion, retning og casting"
+
+#: SourceX/DiabloUI/credits_lines.cpp:90
+msgid "Script & Story"
+msgstr "Manuskript og historie"
+
+#: SourceX/DiabloUI/credits_lines.cpp:94
+msgid "Voice Editing"
+msgstr "Stemmedigering"
```
Comment (AJenbo) [inline translations/da.po:260]:
```suggestion
msgstr "Stemmeredigering"
```

File: translations/da.po [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+msgid "Voices"
+msgstr "Stemmer"
+
+#: SourceX/DiabloUI/credits_lines.cpp:102
+msgid "Recording Engineer"
+msgstr "Optagelsesingeniør"
+
+#: SourceX/DiabloUI/credits_lines.cpp:105
+msgid "Manual Design & Layout"
+msgstr "Manuel design og layout"
```
Comment (AJenbo) [inline translations/da.po:272]:
```suggestion
msgstr "Design af manual og layout"
```

File: translations/da.po [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+
+#: SourceX/DiabloUI/credits_lines.cpp:109
+msgid "Manual Artwork"
+msgstr "Manuel kunst"
```
Comment (AJenbo) [inline translations/da.po:276]:
```suggestion
msgstr "Manualkunst"
```

File: translations/da.po [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+msgid "QA Counterintelligence"
+msgstr "QA-modintelligens"
+
+#: SourceX/DiabloUI/credits_lines.cpp:131
+msgid "Order of Network Information Services"
+msgstr "Bestilling af netværksinformationstjenester"
+
+#: SourceX/DiabloUI/credits_lines.cpp:135
+msgid "Customer Support"
+msgstr "Kunde support"
```
Comment (AJenbo) [inline translations/da.po:304]:
```suggestion
msgstr "Kundesupport"
```

File: translations/da.po [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+msgid "Multi Player Game"
+msgstr "Multiplayer-spil"
+
+#: SourceX/DiabloUI/selconn.cpp:39
+msgid "Requirements:"
+msgstr "Krav:"
+
+#: SourceX/DiabloUI/selconn.cpp:41
+msgid "no gateway needed"
+msgstr "ingen gateway"
```
Comment (AJenbo) [inline translations/da.po:443]:
```suggestion
msgstr "gateway behøves ikke"
```
(check if this fits first)

File: translations/da.po [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+msgstr "Vælg handling"
+
+#: SourceX/DiabloUI/selgame.cpp:87
+msgid "Enter Password"
+msgstr "Indtast adgangskode"
+
+#: SourceX/DiabloUI/selgame.cpp:120
+msgid "Create a new game with a difficulty setting of your choice."
+msgstr ""
+"Opret et nyt spil med en indstilling af vanskeligheder efter eget valg."
```
Comment (AJenbo) [inline translations/da.po:509]:
```suggestion
"Opret et nyt spil med sværhedsgrad efter eget valg."
```

File: translations/da.po [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+#: SourceX/DiabloUI/selgame.cpp:140
+msgid "Join TCP Games"
+msgstr "Deltag i et TCP-spil"
+
+#: SourceX/DiabloUI/selgame.cpp:158
+msgid ""
+"Normal Difficulty\n"
+"This is where a starting character should begin the quest to defeat Diablo."
+msgstr ""
+"Normal vanskelighed\n"
```
Comment (AJenbo) [inline translations/da.po:525]:
```suggestion
"Normal sværhedsgrad\n"
```

File: translations/da.po [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+"Det er her en nybegynder skal starte stræben efter at besejre Diablo."
```
Comment (AJenbo) [inline translations/da.po:526]:
```suggestion
"Det er her en nybegynder skal starte sin mission/søgen at besejre Diablo."
```

File: translations/da.po [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+
+#: SourceX/DiabloUI/selgame.cpp:162
+msgid ""
+"Nightmare Difficulty\n"
+"The denizens of the Labyrinth have been bolstered and will prove to be a "
+"greater challenge. This is recommended for experienced characters only."
+msgstr ""
+"Mareridt sværhedsgrad\n"
+"Borgerne i labyrinten er styrket og vil vise sig at være en større "
+"udfordring. Dette anbefales kun til erfarne helt."
```
Comment (AJenbo) [inline translations/da.po:536]:
```suggestion
"udfordring. Dette anbefales kun til erfarne helte."
```

File: translations/da.po [hunk 12]
```diff
@@ -0,0 +1,693 @@
...
+#: SourceX/DiabloUI/selhero.cpp:205 SourceX/DiabloUI/selhero.cpp:254
+msgid "New Single Player Hero"
+msgstr "Ny singleplayer-helt"
+
+#: SourceX/DiabloUI/selhero.cpp:245
+msgid ""
+"The Rogue and Sorcerer are only available in the full retail version of "
+"Diablo. Visit https://www.gog.com/game/diablo to purchase."
+msgstr ""
+"The skytten  og troldmanden  er kun tilgængelig i den fulde detailversion af "
```
Comment (AJenbo) [inline translations/da.po:641]:
```suggestion
"Skytten og Troldmanden er kun tilgængelige i den fulde detailversion af "
```

File: translations/da.po [hunk 13]
```diff
@@ -0,0 +1,693 @@
...
+msgid "Character already exists. Do you want to overwrite \"%s\"?"
+msgstr "Karakteren findes allerede. Vil du overskrive \"%s\"?"
+
+#: SourceX/DiabloUI/selhero.cpp:301
+msgid "Unable to create character."
+msgstr "Kunne ikke oprette karakter."
+
+#: SourceX/DiabloUI/selhero.cpp:382
+msgid "Delete Multi Player Hero"
+msgstr "Set multiplayer-helt"
```
Comment (AJenbo) [inline translations/da.po:663]:
```suggestion
msgstr "Slet multiplayer-helt"
```

**Pull request closed without merging**